### PR TITLE
Preg match regex mutators

### LINKED ISSUE #324: Implement Regex Mutator
@kalessil suggested a set of interesting mutators for Regular Expressions.

```diff
- preg_match("/test/i", $string));
+ preg_match("/test/", $string));

- preg_match("/test/g", $string));
+ preg_match("/test/", $string));

- preg_match("/^test$/", $string));
+ preg_match("/^test/", $string));
+ preg_match("/test$/", $string));
```

and so on.
Original message:

> E.g. add/remove i-flag, `\d,\w\s` <-> `\D,\W\S`, `{<number>[, {number}]}` - same as hardcoded numbers mutations, `.*` <-> `.+`

thoughts?

-----

UPD: now it's clear that we should implement it https://twitter.com/infection_php/status/990136566726189057

### PR DESCRIPTION
**Pull request by adeptofvoltron**
This PR:

- [x] Adds new feature ...
- [x] Covered by tests
- [x] Fixes #324
- [ ] Documentation PR (like at https://github.com/infection/site/pulls)

These mutators works only if in `preg_match` method regex string is passed as an mutator.
to make it work also with variables, or some concat I need to learn more about ATS.

### CODE REVIEW

**Comment by adeptofvoltron** [comment]
@BackEndTea sure. ask what ever you want. As I wrote I'm just trying to learn more about your tool...that why I make contributions for it...I am aware that some of them might not be perfect.
(btw. already created coalesce mutator...probably I will push it in other PR)

**Comment by BackEndTea** [comment]
My bad i never 'finished' the review, but left it on pending by mistake

File: src/Mutator/Regex/AbstractPregMatch.php [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+     */
+    protected function mutatesNode(Node $node): bool
+    {
+        return $node instanceof FuncCall &&
+            $node->name instanceof Node\Name &&
+            strtolower((string) $node->name) == 'preg_match'
+            && $node->args[0]->value instanceof Node\Scalar\String_;
+    }
+
+    protected function pullOutPattern(Node\Arg $argument): string
```
[the remark this answers is not in the document]
Comment (adeptofvoltron) [reply src/Mutator/Regex/AbstractPregMatch.php]: sure. same with setter. tommorow I will handle it

File: src/Mutator/Regex/PregMatchSwapCaret.php [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+namespace Infection\Mutator\Regex;
+
+/**
+ * @internal
+ */
+final class PregMatchSwapCaret extends AbstractPregMatch
+{
+    protected function manipulatePattern(string $pattern): string
+    {
+        preg_match('/^([\/#~+])([\^]?)([^\^]*)\1([gmixXsuUAJD]*)$/', $pattern, $matches);
```
[the remark this answers is not in the document]
Comment (adeptofvoltron) [reply src/Mutator/Regex/PregMatchSwapCaret.php]: hmm there is an warning when encounter invalid regex into preg_match...but you are right...
that regex will require some correction.

**Comment by adeptofvoltron** [comment]
actually I took an idea from here:  https://github.com/infection/infection/issues/324

**Comment by borNfreee** [comment]
I agree with @sanmai, no need to *add* `^` or `$`, it makes sense only to remove them. Could you please update the PR?

**Comment by sanmai** [comment]
I took liberty to update your PR with a documentation requirement. That's super simple, [looks like this](https://github.com/infection/site/pull/44/files).

**Comment by adeptofvoltron** [comment]
sure. sorry for being inactive for last few days...probably I will be back next week (just getting married) :P

**Comment by adeptofvoltron** [comment]
sure. sorry for being inactive for last few days...probably I will be back next week (just getting married) :P

**Comment by BackEndTea** [comment]
@adeptofvoltron, I hope married life is treating you well 😃
Do you still have to time to pursue this PR, or would you like someone to take over this PR for you?

**Comment by adeptofvoltron** [comment]
@BackEndTea indeed is treating me well. But I was a litytle bit busy. Just sitting back to that PR :)

File: tests/Mutator/Regex/PregMatchRemoveCaretTest.php [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+preg_match('~^some-regexp~ig', 'irrelevant');
+PHP
+            ,
+            <<<'PHP'
+<?php
+
+preg_match('~some-regexp~ig', 'irrelevant');
+PHP
+        ];
+    }
```
Comment (sanmai) [inline tests/Mutator/Regex/PregMatchRemoveCaretTest.php]: You need a negative case also, e.g. where it does not mutate.

File: tests/Mutator/Regex/PregMatchRemoveDollarTest.php [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+preg_match('~some-regexp$~ig', 'irrelevant');
+PHP
+            ,
+            <<<'PHP'
+<?php
+
+preg_match('~some-regexp~ig', 'irrelevant');
+PHP
+        ];
```
Comment (sanmai) [inline tests/Mutator/Regex/PregMatchRemoveDollarTest.php]: Same as above.

File: src/Mutator/Regex/AbstractPregMatch.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    {
+        /** @var Node\Scalar\String_ $stringNode */
+        $stringNode = $argument->value;
+
+        return $stringNode->value;
+    }
+
+    abstract protected function manipulatePattern(string $pattern): string;
+
+    abstract protected function isProperRegexToMutate(string $pattern): bool;
```
Comment (sanmai) [inline src/Mutator/Regex/AbstractPregMatch.php]: These functions are being called in the following order: `isProperRegexToMutate`, `pullOutPattern`, `manipulatePattern`, and `setNewPattern` lastly. Yet they're defined in a slightly different order. Am I missing some clues here?

File: src/Mutator/Regex/PregMatchRemoveDollar.php [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+declare(strict_types=1);
+
+namespace Infection\Mutator\Regex;
+
+/**
+ * @internal
+ */
+final class PregMatchRemoveDollar extends AbstractPregMatch
+{
+    public const ANALYSE_REGEX = '/^([\/#~+])([^$]*)([$]?)\1([gmixXsuUAJD]*)$/';
```
Comment (sanmai) [inline src/Mutator/Regex/PregMatchRemoveDollar.php]: So we won't change an expression if someone decides to use a colon as a delimiter?

File: tests/Fixtures/e2e/Trait_Coverage/expected-output.txt [hunk 7]
```diff
@@ -3,9 +3,6 @@ Killed mutants:
 ===============
 
 
-Mutator: PublicVisibility
-Line 10
-
```
Comment (sanmai) [inline tests/Fixtures/e2e/Trait_Coverage/expected-output.txt]: It appears that CI is failing because of this change.

**Comment by sanmai** [comment]
These are fine mutators. I wonder if you have a minute to give them another try.

**Comment by maks-rafalko** [comment]
Continued here: https://github.com/infection/infection/pull/1455

**Pull request closed without merging**